llvm/lib/Target/Primate/PrimateVLIWPacketizer.cpp

```diff
@@ -146,6 +148,66 @@ bool PrimatePacketizer::runOnMachineFunction(MachineFunction &MF) {
   return true;
 }
 
+bool PrimatePacketizerList::insertBypassOps(MachineInstr* br_inst, llvm::SmallVector<MachineInstr*, 6>& generated_bypass_instrs) {
+  
+  for (auto& operand : br_inst->uses()) {
+    // skip non-reg
+    if (!operand.isReg())
+      continue;
+    
+    // check if someone already generates this operand
+    bool operand_generated = false;
+    for (auto& otherMI : CurrentPacketMIs) {
+      // skip the branch subinstruction itself
+      if (otherMI == br_inst)
+        continue;
+      for (auto& otherOperand : otherMI->defs()) {
+        // skip non-reg
+        if (!otherOperand.isReg())
+          continue;
+        // if the reg indices match, the producer has been found
+        if (otherOperand.getReg() == operand.getReg()) {
+          operand_generated = true;
+        }
+      }
+    }
+    if(operand_generated) {
+      LLVM_DEBUG({dbgs() << "PrimatePacketizerList::endPacket found gernerator for "; operand.dump();});
+      continue;
+    }
+
+    // no one generated this operand. attempt to a bypass op.
+    LLVM_DEBUG(dbgs() << "PrimatePacketizerList::endPacket No producer op for branch instr. Attempt bypass op.\n");
+    MachineInstr* bypass_op = BuildMI(*br_inst->getParent(), br_inst, llvm::DebugLoc(), PII->get(Primate::ADDI), operand.getReg())
+              .addReg(operand.getReg())
+              .addImm(0);
+    bool ResourceAvail = ResourceTracker->canReserveResources(*bypass_op);
+    if (!ResourceAvail) {
+      // no room. set up packet for next iterations. 
+      // 1) Need to put all the bypass ops into the BB above the Branch.
+      // 2) Move the end of packet pointer to before the bypass.
+      // 3) end packet as normal WITHOUT the branch 
+      // 4) insert the bypasses into the current packet.
+      // 5) return to main loop
+      
+      // bypass_op->print(dbgs());
+      // LLVM_DEBUG(dbgs() << "============ Current Packet ============\n");
+      // br_inst->getParent()->dump();
+      // LLVM_DEBUG(dbgs() << "============ Bypass Op Basic Block ============\n");
+      // bypass_op->getParent()->dump();
+      LLVM_DEBUG(dbgs() << "PrimatePacketizerList::endPacket cannot insert bypass_instr! no resources!\n");
+      generated_bypass_instrs.push_back(bypass_op);
+      return true; // push the br and bypasses to the next packet.
+    }
+    else {
+      LLVM_DEBUG({dbgs() << "PrimatePacketizerList::endPacket Bypass instr inserted for operand: ";  operand.dump(); });
+      ResourceTracker->reserveResources(*bypass_op);
```

Comment:
We don't actually reserve resources until we know we don't need a new packet.

Reply:
Part of fix in [5c29c54](https://github.com/FAST-Research-Group/primate-arch-gen/pull/1/commits/5c29c542e7c288e70fb7592a57b554905880ebe9)